src/main/java/module/lineup/assistant/LineupAssistant.java

```diff
@@ -53,159 +53,250 @@ public final boolean isPlayerInStartingEleven(int spielerId, Vector<MatchLineupP
+		);
+		var backupCandidates = new ArrayList<>(substitutePositions.stream()
+				.map(MatchLineupPosition::getPlayer)
+				.filter(Objects::nonNull)
+				.toList());
+		lPositions = allPositions.stream().filter(MatchRoleID::isBackupsMatchRoleId).toList();
+		for (var b : backupPositionOrder) {
+			// local backup candidates list without player of corresponding substitutes slot
+			var substitutePosition = getSubstitutePosition(b.get(0));
+			var matchLineupPosition = allPositions.stream().filter(i->i.getRoleId()==substitutePosition).findFirst();
+			if (matchLineupPosition.isPresent()) {
+				var substitutePlayerId = matchLineupPosition.get().getPlayerId();
+				var backupCandidatesForPosition = new ArrayList<>(backupCandidates.stream().filter(i->i.getPlayerId() != substitutePlayerId).toList());
+				optimizeLineup(b, backupCandidatesForPosition, lPositions);
+			}
 		}
 	}
 
-	private void addForward(ArrayList<Byte> fieldPlayerPositionOrder, ArrayList<Byte> reservePositionOrder) {
-		// FORWARD
-		fieldPlayerPositionOrder.add(IMatchRoleID.FORWARD);
-		fieldPlayerPositionOrder.add(IMatchRoleID.FORWARD_DEF);
-		fieldPlayerPositionOrder.add(IMatchRoleID.FORWARD_TOWING);
+	private int getSubstitutePosition(byte position) {
+		return switch (position){
+			case IMatchRoleID.KEEPER -> IMatchRoleID.substGK1;
+			case IMatchRoleID.CENTRAL_DEFENDER -> IMatchRoleID.substCD1;
+			case IMatchRoleID.BACK -> IMatchRoleID.substWB1;
+			case IMatchRoleID.MIDFIELDER -> IMatchRoleID.substIM1;
+			case IMatchRoleID.WINGER -> IMatchRoleID.substWI1;
+			case IMatchRoleID.FORWARD -> IMatchRoleID.substFW1;
+			case IMatchRoleID.EXTRA -> IMatchRoleID.substXT1;
+            default -> throw new IllegalStateException("Unexpected value: " + position);
+        };
+	}
 
-		reservePositionOrder.add(IMatchRoleID.FORWARD);
+	private void optimizeLineup(List<Byte> requestedPositions, ArrayList<Player> players, List<MatchLineupPosition> lineupPositions) {
+		if ((lineupPositions == null) || (players == null)) return;
+		for (var pos : lineupPositions) {
+			if (pos.getPlayerId() == 0 &&   // there isn't already a player at this position
+					requestedPositions.contains(pos.getPosition())) {
+
+				var behaviours = new ArrayList<>(MatchLineupPosition.getBehaviours(pos.getRoleId()));
+				if (behaviours.isEmpty()) {behaviours.add(IMatchRoleID.NORMAL);}
+				var maxRating = -1.0;
+				Player bestPlayer = null;
+				byte bestBehaviour = 0;
+				for (var behaviour : behaviours) {
+					for (var player : players) {
+						var r = player.getMatchBeginningRating(getRatingPosition(pos.getRoleId()), behaviour);
+						if (r > maxRating) {
+							maxRating = r;
+							bestPlayer = player;
+							bestBehaviour = behaviour;
+						}
+					}
+				}
+
+				// fill the position
+				if (bestPlayer != null) {
+					pos.setPlayerIdIfValidForLineup(bestPlayer.getPlayerId());
+					pos.setBehaviour(bestBehaviour);
+					if (!pos.isBackupsMatchRoleId()) {
+						players.remove(bestPlayer);
```

Comment:
One thing that troubles me, and you're probably better placed here to comment, is that we seem to give precedence to the order in which we find players in the list, which may not be the optimal one (another player further down the list could  be better in an earlier position, and the resulting line-up could overall be better).  Is this a a fair thing to say?  

Obviously exploring multiple orders would be expensive from a resource perspective, but I wonder if that could be optimised somehow.

Reply:
Another thing I notice is that the attitude selection maximises the player rating at a given position, but doesn't necessarily maximise the expected sector: I have an example of a winger that has a higher rating as defensive, but when set to normal increases midfield slightly, and decreases defence a fair bit.  This makes total sense, but I wonder this is a behaviour the user would expect?

There could be a separate option in the future that aims to maximise selected sector (maybe based on the choice AW/MF/ST?), but probably not relevant at this point.

Reply:
The order of the player list has no impact to the optimized lineup - only the order of the positions/sectors.

The calculated player rating is quite similar to the hatstat calculation. I fear it would be very hard to find further improvements. With the help of the sector preference setting, the desired sector is already preferred.